### PR TITLE
Provide parameter to dask.array.gufunc to fix output dtype bug in np.vectorize

### PR DESCRIPTION
**Pull request by ryan-deak-zefr**
`np.vectorize` can mangle the desired dtypes that are passed to it when creating a vectorized wrapper of a function.  This can be fixed by reassigning the output dtypes passed to `np.vectorize` to the wrapped function returned by `np.vectorize`.  

Because `apply_gufunc` calls `np.vectorize` inside its body, this fix cannot be applied in the dask gufunc case.  To overcome this, a Boolean keyword argument `reassign_vectorize_dtypes` is introduced that when provided and true, causes then reassignment of the output dtypes to the function wrapped by `np.vectorize`.  This provides enough power to fix the issue while not changing the default behavior.

- [ ] Tests added / passed
- [ ] Passes `flake8 dask`

### CODE REVIEW

File: dask/array/gufunc.py [hunk 1]
```diff
@@ -292,6 +301,14 @@ def apply_gufunc(func, signature, *args, **kwargs):
     if vectorize:
         func = np.vectorize(func, signature=signature, otypes=otypes)
 
+        # This overcomes a bug in ``np.vectorize`` whereby unicode dtypes with
+        # prescribed lengths are mangled.  E.g., without reassigning output
+        # dtypes, a desired output dtype of "U2" will becomes "U1" in
+        # ``np.vectorize``.  To overcome this, simply reset the output dtypes
+        # to the ones supplied to ``np.vectorize``.
```
Comment (jcrist) [inline dask/array/gufunc.py]: Is there an upstream issue in numpy that we could reference here?

File: dask/array/gufunc.py [hunk 2]
```diff
@@ -292,6 +301,14 @@ def apply_gufunc(func, signature, *args, **kwargs):
...
+        if reassign_vectorize_dtypes:
+            func.otypes = otypes
```
Comment (jcrist) [inline dask/array/gufunc.py]: For reference, the following demonstrates the issue and the fix:

```python
In [1]: import numpy as np

In [2]: x = np.array([['12', '23'], ['34', '45']])

In [3]: def identity(x):
   ...:     return x
   ...:

In [4]: func = np.vectorize(identity, signature='()->()', otypes=('U2',))

In [5]: func(x)  # output is incorrectly trimmed
Out[5]:
array([['1', '2'],
       ['3', '4']], dtype='<U1')

In [6]: func.otypes = ('U2',)

In [7]: func(x)  # output is now correct
Out[7]:
array([['12', '23'],
       ['34', '45']], dtype='<U2')

In [8]: func = np.vectorize(identity, otypes=('U2',))

In [9]: func(x)  # When no signature provided, everything works fine
Out[9]:
array([['12', '23'],
       ['34', '45']], dtype='<U2')
```

File: dask/array/gufunc.py [hunk 3]
```diff
@@ -212,6 +212,13 @@ def apply_gufunc(func, signature, *args, **kwargs):
     vectorize: bool, keyword only
         If set to ``True``, ``np.vectorize`` is applied to ``func`` for
         convenience. Defaults to ``False``.
+    reassign_vectorize_dtypes: bool, keyword only
+        If set to ``True``, after ``np.vectorize`` is applied to ``func``,
+        then the output dtypes are assigned to func.otypes.  This overcomes
+        a bug in numpy.vectorize that mangles dtypes.  This only matters if
+        ``vectorize`` is ``True``.  This should most likely be set to ``True``
+        but is defaults to ``False`` for backward compatibility.  Defaults
+        to ``False``.
```
Comment (jcrist) [inline dask/array/gufunc.py]: There is a ux cost to adding new options to methods. I suspect that no users are relying on the current broken behavior, and we could apply this fix universally with no option (my preference). Thoughts?

File: dask/array/tests/test_gufunc.py [hunk 4]
```diff
@@ -541,3 +541,49 @@ def mysum(x, res):
...
+        f = da.gufunc(identity,
+                      output_dtypes=(o_dtype,),
+                      signature="()->()",
+                      vectorize=True)
+    else:
+        f = da.gufunc(identity,
+                      output_dtypes=(o_dtype,),
+                      signature="()->()",
+                      vectorize=True,
+                      reassign_vectorize_dtypes=reassign_vectorize_dtypes)
```
Comment (jcrist) [inline dask/array/tests/test_gufunc.py]: It would be good to test with a few different dtypes and signatures (perhaps one outputting multiple arrays).

**Comment by mrocklin** [comment]
@ryan-deak-zefr any update here?  @jcrist posted some questions and comments above.

**Comment by mrocklin** [comment]
@ryan-deak-zefr checking in again.  Do you plan to return to this?  Otherwise I'm likely to close it soon.

**Comment by jrbourbeau** [comment]
Closing due to lack of response. Please feel free to reopen if you'd like to pick this PR up again.

**Pull request closed without merging**